framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+        writer.writeImpl(scene, new PrintWriter(out));
+    }
+
+    /**
+     * Writes the annotations in {@code scene} to the file {@code filename} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param basetypes a map from the {@code description} field of {@code ATypeElement}s to the
+     *     {@code TypeMirror}s that represent their base Java types
+     * @param types a map from fully-qualified names to the {@code ATypeElement}s representing their
+     *     declarations
+     * @param enumNamesToEnumConstant a map from fully-qualified enum names to the enum constants
+     *     defined in that name
+     * @param filename the path of the file to write to
+     * @throws IOException if the file doesn't exist
+     * @see #write(AScene, Map, Map, Map, Writer)
+     */
+    public static void write(
+            AScene scene,
+            Map<String, TypeMirror> basetypes,
+            Map<@FullyQualifiedName String, TypeElement> types,
+            Map<@FullyQualifiedName String, List<VariableElement>> enumNamesToEnumConstant,
+            String filename)
+            throws IOException {
+        write(scene, basetypes, types, enumNamesToEnumConstant, new FileWriter(filename));
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the package.
+     *
+     * @param className the binary name of a class
+     * @return the part of the name referring to the package
+     */
+    private static String packagePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? "" : className.substring(0, lastdot);
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the basename of the class. This method
+     * replaces the {@code $}s in the names of inner classes with {@code .}s, so that they can be
+     * printed correctly in stub files.
+     *
+     * @param className a binary name
+     * @return the part of the name representing the class's name without its package
+     */
+    private static String basenamePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        String result = (lastdot == -1) ? className : className.substring(lastdot + 1);
+        return result.replace('$', '.');
+    }
+
+    /**
+     * Formats a literal argument of an annotation. Similar to {@link IndexFileWriter#printValue},
+     * but does not print directly. Instead, returns the result to be printed.
+     *
+     * @param aft the annotation whose values are being formatted, for context
+     * @param o the value or values to format
+     * @return the String representation of the value
+     */
+    private String formatAnnotationValue(AnnotationFieldType aft, Object o) {
+        if (aft instanceof AnnotationAFT) {
+            return formatAnnotation((Annotation) o);
+        } else if (aft instanceof ArrayAFT) {
+            StringBuilder result = new StringBuilder();
```

Comment:
Rather than creating a StringBuilder here and a StringJoiner below, just create a StringJoiner here (using the 3-argument form).  That reduces code size and improves efficiency.